### PR TITLE
Block processing eip4844

### PR DESCRIPTION
**Pull request by realbigsean**
## Issue Addressed

4844 blob processing: https://github.com/ethereum/consensus-specs/blob/dev/specs/eip4844/beacon-chain.md

@michaelsproul  wondering if you could check this out - let me know if you think I went overboard with the iterators. It makes it looks less like the spec and makes the error handling a bit trickier

### CODE REVIEW

File: consensus/state_processing/src/per_block_processing/eip4844/eip4844.rs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+use types::consts::eip4844::{BLOB_TX_TYPE, VERSIONED_HASH_VERSION_KZG};
+use types::{
+    AbstractExecPayload, BeaconBlockBodyRef, EthSpec, ExecPayload, FullPayload, FullPayloadRef,
+    KzgCommitment, Transaction, Transactions, VersionedHash,
+};
+
+pub fn process_blob_kzg_commitments<T: EthSpec, Payload: AbstractExecPayload<T>>(
+    block_body: BeaconBlockBodyRef<T, Payload>,
+) -> Result<(), BlockProcessingError> {
+    if let (Ok(payload), Ok(kzg_commitments)) = (
```
Comment (michaelsproul) [inline consensus/state_processing/src/per_block_processing/eip4844/eip4844.rs:16]: We can probably `?` these away right? Once 4844 is activated we can expect both `payload` + `blob_kzg_commitments` can't we?

File: consensus/state_processing/src/per_block_processing/eip4844/eip4844.rs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+/// Only transactions of type `BLOB_TX_TYPE` should be passed into this function.
+fn tx_peek_blob_versioned_hashes<T: EthSpec>(
+    opaque_tx: &Transaction<T::MaxBytesPerTransaction>,
+) -> Result<
+    impl IntoIterator<Item = Result<VersionedHash, BlockProcessingError>> + '_,
+    BlockProcessingError,
+> {
+    let tx_len = opaque_tx.len();
+    let message_offset = 1 + u32::from_ssz_bytes(opaque_tx.get(1..5).ok_or(
```
Comment (michaelsproul) [inline consensus/state_processing/src/per_block_processing/eip4844/eip4844.rs]: my spidey senses are tingling for all this arithmetic not using `safe_arith`, but we can patch that up later

**Review by michaelsproul — APPROVED** [review]
the iterators look good!

File: consensus/state_processing/src/per_block_processing/eip4844/eip4844.rs (hunk 1 above)
Comment (realbigsean) [reply consensus/state_processing/src/per_block_processing/eip4844/eip4844.rs:16]: There's no check elsewhere in block processing before this that we're in 4844 so if we error here we couldn't process pre 4844 blocks - but this raises the question of how we should be handling fork specific logic in the `per_block_processing` method generally cause it isn't obvious that it's shared and what logic exactly is shared across what fork. Maybe the current set up of "process this field if it exists" is the best solution but I'm not sure

File: consensus/state_processing/src/per_block_processing/eip4844/eip4844.rs (hunk 2 above)
Comment (realbigsean) [reply consensus/state_processing/src/per_block_processing/eip4844/eip4844.rs]: good call, added it here: 7f54ab8a5b9d9d96de07b7f2ef63fb8659670ba1

**Comment by realbigsean** [comment]
Thanks for checking it out michael! I'm gonna merge to the 4844 to move us closer to having something testable

I snuck in the change here FYI https://github.com/ethereum/consensus-specs/pull/3027

**Pull request merged**